Documentation/general/cli-ux-guidelines.md

```diff
@@ -0,0 +1,155 @@
+    mnemonic.
+
+    -   **Example:** “dotnet build -o path/to/where/to/put/output”
+
+-   The long format starts with double dashes (“--“) and can have one or more
+    words. Multiple words are separated by a single dash (“-“).
+
+    -   **Example:** “dotnet test --test-case-filter”
+
+-   The double dashes (“--“) on their own mark the **argument separator**.
+    Anything after this option is passed verbatim to whatever the command starts
+    as a sub-process.
+
+    -   **Example:** “dotnet run -- --app-arg-1”
+
+-   Windows-style forward slash options (e.g. “/p”) are supported **only for
+    MSBuild parameters.**
+
+-   PowerShell-style single dashes with a work (e.g. “-Filter”) are **not
+    supported**.
+
+-   Help options are predefined as “-h \| --help” and those should be used.
+
+There are common options in the CLI that are reserved for certain concepts. The
+table below outlines those common options. If a command needs to accept the same
+options, it should use these and it must not override their semantics (e.g. use
+“--output” to mean something different or redefine what “-f” means for just that
+command).
+
+| Long option      | Short option | Description                                                  |
+|------------------|--------------|--------------------------------------------------------------|
+| \--framework     | \-f          | Specifies which framework (target) to use in the invocation. |
+| \--output        | \-o          | Specifies where the output of the command should be placed.  |
+| \--runtime       | \-r          | Specifies the runtime ID (RID) to use in the invocation.     |
+| \--configuration | \-c          | Specifies the configuration to use in the invocation.        |
+| \--verbosity     | \-v          | Specifies the verbosity level of the command.                |
+| \--help          | \-h          | Shows the help for the command.                              |
+
+One special area is interacting with MSBuild. If a command has the need to do
+that, there are two additional requirements with regards to options:
+
+1.  The commands need to be able to take “/p” parameters in addition to whatever
+    other properties they require.
+
+2.  If the command invokes a predefined MSBuild target, it needs to block
+    “/t:\<target\>” and “/target:\<target\>” options and throw an error that is
+    pre-defined in the CLI.
+
+It is important to note that commands that want to invoke user-specified targets
+should not be submitted to the CLI, since the CLI comes with “dotnet msbuild”
+command that does precisely that.
+
+## Arguments
+Arguments can be named however the commands needs them. We do have predefined
+argument names for the SLN file and project file. They are defined in the CLI
+source code and should be used if the command has the need to use those two
+arguments.
+
+## Help messages
+Help messages are automatically generated based on the arguments, options and

```

Comment:
It is not really implementation, it is mostly to say that people who want to PR and add a new command so that they know they should not need to do anything. 

We don't have guidelines for help messages.